src/transformers/generation/utils.py

```diff
@@ -3365,34 +3394,92 @@ def _beam_search(
             `return_dict_in_generate=True` or a [`~generation.GenerateBeamEncoderDecoderOutput`] if
             `model.config.is_encoder_decoder=True`.
         """
-        # init values
+
+        # 1. Auxiliary tensor manipulation functions
+        def flatten_beam_dim(tensor: torch.Tensor) -> torch.Tensor:
+            """[batch_size, num_beams, ...] -> [batch_size * num_beams, ...]"""
+            shape = list(tensor.shape)
+            return torch.reshape(tensor, [shape[0] * shape[1]] + shape[2:])
+
+        def unflatten_beam_dim(tensor: torch.Tensor, batch_size: int, num_beams: int) -> torch.Tensor:
+            """[batch_size * num_beams, ...] -> [batch_size, num_beams, ...]"""
+            shape = list(tensor.shape)
+            return torch.reshape(tensor, [batch_size, num_beams] + shape[1:])
+
+        def gather_beams(tensor: torch.Tensor, beam_indices: torch.Tensor) -> torch.Tensor:
+            """
+            Gathers the beam slices indexed by beam_indices into new beam array.
+
+            Args:
+                tensor (`torch.Tensor`): A tensor containing data to be gathered. The tensor is a 2D or a 3D tensor
+                    with the two first dimensions depicting the batch and the beam dimensions.
+                beam_indices (`torch.Tensor` of shape `(batch_size, num_beams_to_select)`): The indices of the beams to
+                    select .
+
+            Returns:
+                A tensor with the selected beams
+            """
+            # `take_along_dim` requires its indices arg to have the same number of dims as `input`
+            while len(beam_indices.shape) < len(tensor.shape):
+                beam_indices = beam_indices.unsqueeze(-1)
+            gathered_tensor = torch.take_along_dim(input=tensor, indices=beam_indices, dim=1)
+            return gathered_tensor
```

Comment:
again please let's avoid nested function, these should just be outside the context of the function